File: frontend/src/types/session.ts
export type SessionStatus = 'pending' | 'generating' | 'complete' | 'failed';
export type Priority = 'P1' | 'P2' | 'P3';
export type TestSection = 'positive' | 'negative' | 'boundary' | 'edge' | 'permission';

export interface QAContext {
  product: string;
  platform: string;
  users_roles: string;
  rules_constraints: string;
  risks: string;
  environment: string;
}

export interface TestCase {
  test_id: string;
  reference_id: string;
  scenario: string;
  requirement_ref: string;
  preconditions: string;
  steps: string;
  test_data: string;
  expected_result: string;
  priority: Priority;
  notes: string;
}

export interface TestCases {
  positive: TestCase[];
  negative: TestCase[];
  boundary: TestCase[];
  edge: TestCase[];
  permission: TestCase[];
}

export interface ScenarioMap {
  main_flows: string[];
  alternate_flows: string[];
  error_flows: string[];
  permission_flows: string[];
  integration_flows: string[];
}

export interface ChecklistResult {
  all_items_traced: string;
  positive_negative_coverage: string;
  boundary_coverage: string;
  edge_coverage: string;
  permission_coverage: string;
  integration_coverage: string;
  no_duplicates: string;
  clear_expected_results: string;
  no_hallucinated_rules: string;
}

export interface QAOutput {
  id: number;
  session_id: number;
  scenario_map: ScenarioMap;
  test_cases: TestCases;
  assumptions: string[];
  questions: string[];
  checklist_result: ChecklistResult;
  checklist: ChecklistResult;
  created_at: string;
}

export interface QASession {
  id: number;
  name: string;
  requirement: string;
  context: QAContext;
  template: string;
  status: SessionStatus;
  created_at: string;
  updated_at: string;
  output?: QAOutput;
}

export interface QASessionListItem {
  id: number;
  name: string;
  requirement: string;
  status: SessionStatus;
  created_at: string;
  updated_at: string;
  test_case_count: number;
}

export interface CreateSessionRequest {
  name: string;
  requirement: string;
  context: QAContext;
  template?: string;
}

export function totalTestCaseCount(testCases: TestCases | null | undefined): number {
  if (!testCases) return 0;
  return (
    (testCases.positive?.length ?? 0) +
    (testCases.negative?.length ?? 0) +
    (testCases.boundary?.length ?? 0) +
    (testCases.edge?.length ?? 0) +
    (testCases.permission?.length ?? 0)
  );
}

File: backend/app/schemas/session.py
from pydantic import BaseModel, Field
from datetime import datetime
from typing import Any
from app.models.session import SessionStatus


class QAContextSchema(BaseModel):
    product: str = ""
    platform: str = ""
    users_roles: str = ""
    rules_constraints: str = ""
    risks: str = ""
    environment: str = ""


class CreateSessionRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    requirement: str = Field(..., min_length=1, max_length=10000)
    context: QAContextSchema
    template: str = "full"


class UpdateSessionRequest(BaseModel):
    name: str | None = None


class TestCaseSchema(BaseModel):
    test_id: str = ""
    reference_id: str = ""
    scenario: str = ""
    requirement_ref: str = ""
    preconditions: str = ""
    steps: str = ""
    test_data: str = ""
    expected_result: str = ""
    priority: str = ""
    notes: str = ""


class TestCasesSchema(BaseModel):
    positive: list[TestCaseSchema] = []
    negative: list[TestCaseSchema] = []
    boundary: list[TestCaseSchema] = []
    edge: list[TestCaseSchema] = []
    permission: list[TestCaseSchema] = []


class ScenarioMapSchema(BaseModel):
    main_flows: list[str] = []
    alternate_flows: list[str] = []
    error_flows: list[str] = []
    permission_flows: list[str] = []
    integration_flows: list[str] = []


class ChecklistResultSchema(BaseModel):
    all_items_traced: str = ""
    positive_negative_coverage: str = ""
    boundary_coverage: str = ""
    edge_coverage: str = ""
    permission_coverage: str = ""
    integration_coverage: str = ""
    no_duplicates: str = ""
    clear_expected_results: str = ""
    no_hallucinated_rules: str = ""


class QAOutputSchema(BaseModel):
    id: int
    session_id: int
    scenario_map: ScenarioMapSchema | None = None
    test_cases: TestCasesSchema | None = None
    assumptions: list[str] = []
    questions: list[str] = []
    checklist_result: ChecklistResultSchema | None = None
    checklist: ChecklistResultSchema | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class QASessionResponse(BaseModel):
    id: int
    name: str
    requirement: str
    context: QAContextSchema | None
    template: str
    status: SessionStatus
    created_at: datetime
    updated_at: datetime | None
    output: QAOutputSchema | None = None

    class Config:
        from_attributes = True


class QASessionListItem(BaseModel):
    id: int
    name: str
    status: SessionStatus
    created_at: datetime
    updated_at: datetime | None
    test_case_count: int = 0

    class Config:
        from_attributes = True
